src/odbc_connection.cpp

```diff
@@ -2782,108 +2843,340 @@ fetch_and_store
+              return_code =
+              SQLGetData
+              (
+                data->hstmt,
+                column_index + 1,
+                data->columns[column_index]->bind_type,
+                target_buffer,
+                buffer_size,
+                &string_length_or_indicator
+              );
+              if (!SQL_SUCCEEDED(return_code))
+              {
+                return return_code;
+              } 
+
+              // If the data is null, simply indicate and continue to the next
+              // column
+              if (string_length_or_indicator == SQL_NULL_DATA)
+              {
+                row[column_index].size = SQL_NULL_DATA;
+                continue;
+              }
+              // If the data (+ whatever null terminator, not included in
+              // string_length_or_indicator) was larger than the size of the
+              // buffer, then need to call SQLGetData at least once more
+              else if (
+                string_length_or_indicator == SQL_NO_TOTAL || 
+                string_length_or_indicator >= buffer_size
+              ) 
+              {
+                // Continue looping as long as we don't know the final resize
+                // that we need. Once a buffer size is returned instead of
+                // SQL_NO_TOTAL, we still need to call SQLGetData one more time,
+                // and the break out of this loop.
+                bool last_get_data_call = false;
+                while (!last_get_data_call)
                 {
-                  SQLCHAR *memory_start = (SQLCHAR *)data->bound_columns[column_index].buffer + (row_index * data->columns[column_index]->buffer_size);
-                  row[column_index].size = strlen((const char *)memory_start);
-                  // Although fields going from SQL_C_CHAR to Napi::String use
-                  // row[column_index].size, NUMERIC data uses atof() which requires
-                  // a null terminator. Need to add an aditional byte.
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size + 1]();
-                  memcpy
+                  // If not SQL_NO_TOTAL, we know how large to resize to buffer
+                  // to get the rest of the data.
+                  if (string_length_or_indicator != SQL_NO_TOTAL)
+                  {
+                    last_get_data_call = true;
+                  }
+
+                  // Handling the data that was returned depends heavily on the
+                  // target type of the binding data.
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_returned_length = buffer_size;
+                      buffer_size =
+                        string_length_or_indicator == SQL_NO_TOTAL ?
+                        buffer_size * 2 :
+                        row[column_index].size + string_length_or_indicator;
```

Comment:
If we got a valid length, the returned value includes the current size, so there's no need to add them.

Reply:
If several calls to `SQLGetData` have returned `SQL_NO_TOTAL`, then the size might be something like 1000, while the last call to `SQLGetData` may have returned 10 (per this comment https://github.com/markdirish/node-odbc/pull/161#discussion_r588010331). In that case, the size of the buffer I need in the end is "everything that was returned prior" (stored in `row[column_index].size` in bytes) + "what was returned in the last call to `SQLGetData` + the rest" (stored in `string_length_or_indicator`).

I do think I am missing some logic if the final call after some `SQL_NO_TOTAL` calls already fits in my existing buffer, in which case I have no reason to `realloc` again. I think the assignment to `data_returned_length` needs to be fixed one line above this snippet to be:

```
data_returned_length = (
  row[column_index].size + string_length_or_indicator <= buffer_size ? 
  string_length_or_indicator : 
  buffer_size - row[column_index].size
);
```
Shouldn't need the same logic for `SQL_C_CHAR` and `SQL_W_CHAR` since the amount of data returned is determined with their `strlen` functions.

Reply:
Ah yeah, that makes sense. Looks like you haven't made the change mentioned. Was that intentional? I don't think it's _necessary_ since the realloc for smaller buffer IIRC isn't going to do anything but return the original pointer.

Reply:
Should be in the latest commit now.